### PR TITLE
move-between-bars

### PR DESCRIPTION
**Pull request by yutytuty**

### CODE REVIEW

File: src/main/java/frc/robot/Commands/MoveBetweenBars.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    private final WPI_TalonSRX leftTalon;
+    private final WPI_TalonSRX rightTalon;
+
+    public MoveBetweenBars(ClimberPlacer dynamicClimber) {
+        super(dynamicClimber, MOVEMENT_SPEED);
+        this.leftTalon = dynamicClimber.getLeftTalon();
+        this.rightTalon = dynamicClimber.getRightTalon();
+    }
+
+    /**
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Commands/MoveBetweenBars.java:33]: remove redundant doc

**Review by tal-sitton — CHANGES_REQUESTED** [review]

**Comment by omri00** [comment]
Ok it's just an idea, tell me what you think.
How about making each side a different insatnce of the same subsystem that only saves 1 motor. This way the logic in the execute moves to `can_move` of the subsystem and you don't need a special command, only a commandGroup with 2 parallel moves

File: src/main/java/frc/robot/Commands/MoveBetweenBars.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    public MoveBetweenBars(ClimberPlacer dynamicClimber) {
+        super(dynamicClimber, MOVEMENT_SPEED);
+        this.leftTalon = dynamicClimber.getLeftTalon();
+        this.rightTalon = dynamicClimber.getRightTalon();
+    }
+
+    @Override
+    public void execute() {
+        if (leftTalon.getStatorCurrent() < STALL_CURRENT.get()) {
```
Comment (omri00) [inline src/main/java/frc/robot/Commands/MoveBetweenBars.java:40]: This whole logic should be in the subsystem, these are hard limits on the system

File: src/main/java/frc/robot/Commands/MoveBetweenBars.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            leftTalon.set(MOVEMENT_SPEED.get());
+        } else {
+            if (!leftPassedFirstStall) {
+                leftPassedFirstStall = true;
```
Comment (omri00) [inline src/main/java/frc/robot/Commands/MoveBetweenBars.java:44]: You never want to set the speed of a motor directly. This should always be thought the subsystem

**Comment by omri00** [comment]
> Ok it's just an idea, tell me what you think. How about making each side a different insatnce of the same subsystem that only saves 1 motor. This way the logic in the execute moves to `can_move` of the subsystem and you don't need a special command, only a commandGroup with 2 parallel moves



> Ok it's just an idea, tell me what you think. How about making each side a different insatnce of the same subsystem that only saves 1 motor. This way the logic in the execute moves to `can_move` of the subsystem and you don't need a special command, only a commandGroup with 2 parallel moves.

Another option is to treat this like a drivetrain with `can_move` and `apply` functions for each side and `move` function that call them both.
You can override the normal `can_move` to return true if one or neither limits are met so you don't have to change the `is_finished` of the command

File: src/main/java/frc/robot/Commands/MoveBetweenBars.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        this.leftTalon = dynamicClimber.getLeftTalon();
+        this.rightTalon = dynamicClimber.getRightTalon();
+    }
+
+    @Override
+    public void execute() {
+        if (leftTalon.getStatorCurrent() < STALL_CURRENT.get()) {
+            leftTalon.set(MOVEMENT_SPEED.get());
+        } else {
+            if (!leftPassedFirstStall) {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/Commands/MoveBetweenBars.java:43]: use the "else if" keyword, it's so much nicer

**Comment by omri00** [comment]
Merge master to here. You have conflicts :(

File: src/main/java/frc/robot/Subsystems/ClimberWinch.java [hunk 5]
```diff
@@ -1,72 +1,47 @@
...
-     */
-    private Level magnetLevel;
-
-    private enum Level {UPPER, LOWER, MIDDLE}
+    private ClimberWinch(Supplier<Double> minSpeed, Supplier<Double> maxSpeed, String namespaceName, MotorController... motorControllers) {
+        super(minSpeed, maxSpeed, namespaceName, motorControllers);
+    }
 
     public static ClimberWinch getInstance() {
         if (instance == null) {
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/ClimberWinch.java:25]: this is not how we write ghost classes

File: src/main/java/frc/robot/Subsystems/ClimberPlacer.java [hunk 6]
```diff
@@ -1,50 +1,38 @@
...
+    private static ClimberPlacer leftInstance, rightInstance;
 
-    public static ClimberPlacer getInstance() {
-        if (instance == null) {
-            leftPlacer = new WPI_TalonSRX(RobotMap.CAN.PLACER_TALON_LEFT);
-            rightPlacer = new WPI_TalonSRX(RobotMap.CAN.PLACER_TALON_RIGHT);
-            instance = new ClimberPlacer(leftPlacer, rightPlacer);
+    public static ClimberPlacer getLeftInstance() {
+        if (leftInstance == null) {
+            leftInstance = new ClimberPlacer(() -> 0.0, () -> 0.0, "", (MotorController) null);
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/ClimberPlacer.java:19]: this is not how we wrtie ghost classes

File: src/main/java/frc/robot/Subsystems/ClimberPlacer.java [hunk 7]
```diff
@@ -1,50 +1,38 @@
...
-        this.frontLimit = new DigitalInput(RobotMap.DIO.PLACER_LIMIT_FRONT);
-        this.backLimit = new DigitalInput(RobotMap.DIO.PLACER_LIMIT_BACK);
-    }
-
-    @Override
-    public boolean canMove(double speed) {
-        return !(ClimberWinch.getInstance().isHooked() || speed > 0 && frontLimit.get() || speed < 0 && backLimit.get());
+    public static ClimberPlacer getRightInstance() {
+        if (rightInstance == null) {
+            rightInstance = new ClimberPlacer(() -> 0.0, () -> 0.0, "", (MotorController) null);
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Subsystems/ClimberPlacer.java:26]: ditto

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Commands/CloseTelescopic.java [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package frc.robot.Commands;
+
+import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
+
+public class CloseTelescopic extends SequentialCommandGroup {
+    public CloseTelescopic() {
+
```
Comment (omri00) [inline src/main/java/frc/robot/Commands/CloseTelescopic.java:7]: ?

File: src/main/java/frc/robot/RobotMap.java [hunk 9]
```diff
@@ -13,8 +13,11 @@
     public interface DIO {
 
         int WINCH_HALL_EFFECT = -1;
-        int PLACER_LIMIT_FRONT = -1;
-        int PLACER_LIMIT_BACK = -1;
-        int PLACER_LIMIT_HOOK = -1;
+        int PLACER_LEFT_LIMIT_FRONT = -1;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/RobotMap.java:16]: what placer? climber placer? intake placer?

File: src/main/java/frc/robot/Commands/DropPlacers.java [hunk 10]
```diff
@@ -0,0 +1,16 @@
+package frc.robot.Commands;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Commands/DropPlacers.java:1]: that's just confusing and unnecessary,
please remove

File: src/main/java/frc/robot/Commands/DropPlacer.java [hunk 11]
```diff
@@ -0,0 +1,27 @@
+package frc.robot.Commands;
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Commands/DropPlacer.java:1]: why do you need to create a class for it? you can just use `MoveGenericSubsystem` and override the `isFinished()` method

File: src/main/java/frc/robot/Commands/DropPlacer.java [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+public class DropPlacer extends MoveGenericSubsystem {
+
+    private final ClimberPlacer placer;
+
+    private final double startTime;
+
+    public DropPlacer(ClimberPlacer placer) {
+        super(placer, placer.getDropSpeed());
+        this.placer = placer;
+        this.startTime = Timer.getFPGATimestamp();
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Commands/DropPlacer.java:16]: the doesn't necessarily start **right** after you create it.

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Commands/DropPlacer.java (hunk 11 above)
Comment (omri00) [reply src/main/java/frc/robot/Commands/DropPlacer.java:1]: And init? I think I deserves it own class

File: src/main/java/frc/robot/Commands/DropPlacers.java (hunk 10 above)
Comment (omri00) [reply src/main/java/frc/robot/Commands/DropPlacers.java:1]: Highly disagree. You want each command to have a single purpose.

File: src/main/java/frc/robot/Commands/DropPlacer.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+    @Override
+    public void initialize() {
+        startTime = Timer.getFPGATimestamp();
+        super.initialize();
+    }
+
+    @Override
+    public boolean isFinished() {
+        return super.isFinished() && placer.isStalling() && initialTimePassed();
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Commands/DropPlacer.java:36]: if one of the limits of the placer is pushed (the placer cant move), you should stop immediately

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Commands/DropPlacers.java (hunk 10 above)
Comment (tal-sitton) [reply src/main/java/frc/robot/Commands/DropPlacers.java:1]: you should K.I.S.S, making a class instead of writing 3 lines of code is redundant.

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Commands/DropPlacers.java [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+
+import edu.wpi.first.wpilibj2.command.ParallelCommandGroup;
+import frc.robot.Subsystems.ClimberPlacer;
+
+public class DropPlacers extends ParallelCommandGroup {
```
Comment (omri00) [inline src/main/java/frc/robot/Commands/DropPlacers.java:6]: Maybe change this to `DropBothPlacers` of something if you've already have dropPlacer

File: src/main/java/frc/robot/Commands/MoveBetweenBars.java [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+import frc.robot.Subsystems.ClimberWinch;
+
+/**
+ * Moves the dynamic climbers downward towards the ground until they hit the third bar.
+ */
+public class MoveBetweenBars extends SequentialCommandGroup {
+
+    public MoveBetweenBars() {
+        ClimberWinch climberWinch = ClimberWinch.getInstance();
+        ClimberPlacer leftClimberPlacer = ClimberPlacer.getLeftInstance();
```
[the remark this answers is not in the document]
Comment (yutytuty) [reply src/main/java/frc/robot/Commands/MoveBetweenBars.java:16]: Requirements

File: src/main/java/frc/robot/Commands/MoveBetweenBars.java [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+        ClimberPlacer rightClimberPlacer = ClimberPlacer.getRightInstance();
```
[the remark this answers is not in the document]
Comment (yutytuty) [reply src/main/java/frc/robot/Commands/MoveBetweenBars.java:17]: Requirements

File: src/main/java/frc/robot/Commands/MoveBetweenBars.java [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+        addRequirements(climberWinch, leftClimberPlacer, rightClimberPlacer);
```
[the remark this answers is not in the document]
Comment (yutytuty) [reply src/main/java/frc/robot/Commands/MoveBetweenBars.java:18]: Shouldn't they be here to make sure they are not used in another command?

[the remark this answers is not in the document]
Comment (tal-sitton) [reply src/main/java/frc/robot/Commands/MoveBetweenBars.java:18]: They **ARE** used by another command, the command that you created. `DropPlacers`

File: src/main/java/frc/robot/Commands/MoveBetweenBars.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (tal-sitton) [reply src/main/java/frc/robot/Commands/MoveBetweenBars.java:16]: the `MoveBetweenBars` doesn't use those, the `DropPlacers` uses them

File: src/main/java/frc/robot/Commands/MoveBetweenBars.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (tal-sitton) [reply src/main/java/frc/robot/Commands/MoveBetweenBars.java:17]: the `MoveBetweenBars` doesn't use those, the `DropPlacers` uses them

File: src/main/java/frc/robot/Commands/DropBothPlacers.java [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+import edu.wpi.first.wpilibj2.command.ParallelCommandGroup;
+import frc.robot.Subsystems.ClimberPlacer;
+
+public class DropBothPlacers extends ParallelCommandGroup {
+
+    public DropBothPlacers() {
+        ClimberPlacer leftPlacer = ClimberPlacer.getLeftInstance();
+        ClimberPlacer rightPlacer = ClimberPlacer.getRightInstance();
+        addRequirements(
+                leftPlacer,
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Commands/DropBothPlacers.java:12]: why new line?

File: src/main/java/frc/robot/Commands/DropPlacer.java [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    private boolean initialTimePassed() {
+        return Timer.getFPGATimestamp() - startTime > initialWaitTime.get();
+    }
+
+    @Override
+    public void initialize() {
+        startTime = Timer.getFPGATimestamp();
+        super.initialize();
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Commands/DropPlacer.java:31]: why `super`?

File: src/main/java/frc/robot/Commands/DropPlacer.java [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+    public static final Supplier<Double> initialWaitTime = rootNamespace.addConstantDouble("initial wait time", 0);
+
+    private double startTime;
+
+    public DropPlacer(ClimberPlacer placer) {
+        super(placer, placer.getDropSpeed());
+        this.placer = placer;
+    }
+
+    private boolean initialTimePassed() {
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Commands/DropPlacer.java:24]: move this to the end of the class

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Commands/DropPlacer.java (hunk 13 above)
Comment (yutytuty) [reply src/main/java/frc/robot/Commands/DropPlacer.java:36]: return !(frontLimit.get() && speed < 0) && !(backLimit.get() && speed > 0); is in the canMove() of the climber placer. This should solve the problem.

File: src/main/java/frc/robot/Subsystems/ClimberPlacer.java [hunk 21]
```diff
@@ -1,50 +1,32 @@
...
-        this.frontLimit = new DigitalInput(RobotMap.DIO.PLACER_LIMIT_FRONT);
-        this.backLimit = new DigitalInput(RobotMap.DIO.PLACER_LIMIT_BACK);
+    public static ClimberPlacer getRightInstance() {
+        return null;
     }
 
-    @Override
-    public boolean canMove(double speed) {
-        return !(ClimberWinch.getInstance().isHooked() || speed > 0 && frontLimit.get() || speed < 0 && backLimit.get());
+    public ClimberPlacer(Supplier<Double> minSpeed, Supplier<Double> maxSpeed, String namespaceName, MotorController... motorControllers) {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/Subsystems/ClimberPlacer.java:21]: constructor should go above getInstance(), which should also go below isStalling() (convention A.III.a.3)

**Review by Yoel-Perman-Brilliant — COMMENTED** [review]
Excellent code

File: src/main/java/frc/robot/Commands/DropPlacer.java (hunk 13 above)
Comment (tal-sitton) [reply src/main/java/frc/robot/Commands/DropPlacer.java:31]: so put || instead of &&?

**Review by tal-sitton — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Commands/DropPlacer.java (hunk 13 above)
Comment (yutytuty) [reply src/main/java/frc/robot/Commands/DropPlacer.java:31]: no

File: src/main/java/frc/robot/Commands/MoveBetweenBars.java [hunk 22]
```diff
@@ -0,0 +1,21 @@
+package frc.robot.Commands;
+
+import com.spikes2212.command.genericsubsystem.commands.MoveGenericSubsystem;
+import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
+import frc.robot.Subsystems.ClimberPlacer;
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/Commands/MoveBetweenBars.java:5]: remove this import

**Review by OfriRosenbaum — DISMISSED** [review]

File: src/main/java/frc/robot/Commands/DropPlacer.java [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+import frc.robot.Subsystems.ClimberPlacer;
+
+import java.util.function.Supplier;
+
+public class DropPlacer extends MoveGenericSubsystem {
+
+    private final ClimberPlacer placer;
+
+    private static final RootNamespace rootNamespace = new RootNamespace("drop placers");
+    public static final Supplier<Double> initialWaitTime = rootNamespace.addConstantDouble("initial wait time", 0);
```
Comment (tal-sitton) [inline src/main/java/frc/robot/Commands/DropPlacer.java:15]: why 0?

**Review by tal-sitton — CHANGES_REQUESTED** [review]

**Review by tal-sitton — APPROVED** [review]

**Review by tal-sitton — APPROVED** [review]

**Pull request merged**